pkg/adaptor/cloud/cloud.go

```diff
@@ -34,9 +35,11 @@ const (
 var logger = log.New(log.Writer(), "[adaptor/cloud] ", log.LstdFlags|log.Lmsgprefix)
 
 type Provider interface {
-	CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator) (instance *Instance, err error)
+	CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator, instanceType string) (instance *Instance, err error)
 	DeleteInstance(ctx context.Context, instanceID string) error
 	Teardown() error
+	// Add SelectInstanceType method based on vCPU and memory to Provider interface
+	SelectInstanceType(ctx context.Context, vCPU int64, memory int64) (instanceType string, err error)
```

Comment:
I suggested to add the  `SelectInstanceType` method, but it turns out that adding the method seems the design complicated.

I think we can directly pass to required instance size information to the `CreateInstance` as follows, instead of adding the `instanceType` argument.

```
CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator, requirement InstanceSpec) (instance *Instance, err error)
```

```
type InstanceSpec struct {
	VCPUs       int
	Memory      int
	MachineType string
}
```

@bpradipt What do you think?

Reply:
I thought about it and felt SelectInstanceType method is helpful. This will give flexibility to implement instance selection logic per provider as needed. 
Now what I can do is passing the entire InstanceTuple or InstanceSpec to the method instead of just cpu/mem which will provide flexibility w.r.to customising the method as needed.

Reply:
I think implementing an instance selection logic per provider is possible with the approach I mentioned above by including `InstanceType` string in the struct.

```
type InstanceSpec struct {
	InstanceType string
	VCPUs        int
	Memory       int
	MachineType  string
}
```

A provider can prepare a private function `selectInstanceType` instead of a public one, and call the function at the beginning of `CreateInstance` like this.

```
instanceType := requirement.InstanceType
if instanceType == "" {
	instanceTypeName = selectInstanceType(requirement.VCPUs, requirement.Memory)
}
```



Another benefit of passing a requirement as a struct data is that we can add additional filed in the future without changing the method signature like this.

```
type InstanceSpec struct {
	InstanceType string
	VCPUs       int
	Memory      int
	MachineType string
	GPUs        int
}
```

Reply:
As I already mentioned, I agree with passing of struct for flexibility and to avoid method signature change again.
Ok I'll remove the public method. Any other comments before I make the changes to limit rework ?

Reply:
Thank you!  I have no other comment for now.

Reply:
Thanks @yoheiueda I updated the code based on your comments. Kindly take a look